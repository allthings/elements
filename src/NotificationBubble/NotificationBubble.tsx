import * as React from 'react'
import View from '../View'
import { css } from 'glamor'
import Icon from '../Icon'
import Text from '../Text'
import { Motion, spring } from 'react-motion'
import Theme from '../Theme'

const styles = {
  bubble: css({
    zIndex: 200,
    position: 'fixed',
    bottom: 50,
    left: '50%',
    transform: 'translate(-50%, 0)',
    margin: '0 auto',
    width: 260,
    height: 68,
    borderRadius: 45,
    boxShadow: '2px 2px 14px 1px rgba(35,35,35,0.2)',
  }),
  text: css({
    width: 200,
    marginLeft: 20,
  }),
}

interface INotificationBubble {
  readonly color?: string
  readonly onTimeout?: () => void
}
interface INotificationBubbleState {
  visible: boolean
}

/**
 * NotificationBubbles are used to inform users about important
 * information, when an action has failed or succeeded. They're a most
 * prominent ways to communicate to merchants. NotificationBubbles are
 * always visible and pop up at the bottom of a page.
 *
 * ## Best practises
 * - Do use them when you otherwise users won't have any feedback.
 * - Don't use them extensive.
 * - Don't use them when no feedback is required.
 *
 * To use NotificationBubbles you need to wrap the **NotificationBubbleManager** around your app.
 * Use `sendSuccess()` exported by the NotificationBubbleManager to show a Notification.
 *
 * ```example
 * import NotificationBubbleManager, { sendSuccess } from '@allthings/elements/behaviour/NotificationBubbleManager'
 * import { Button } from '@allthings/elements'
 *
 * const ShowNotification = () => (
 *   <NotificationBubbleManager>
 *     <View>
 *       <Button onClick={() => sendSuccess('Congratulations, you clicked the Button')}>
 *         Click me!
 *       </Button>
 *     </View>
 *   </NotificationBubbleManager>
 * )
 * ```
 *
 */
class NotificationBubble extends React.Component<
  INotificationBubble,
  INotificationBubbleState
> {
  state = {
    visible: true,
  }

  static defaultProps = {
    color: 'primary',
  }

  componentDidMount() {
    setTimeout(() => this.setState({ visible: false }), 2750)
  }

  handleRest = () =>
    this.state.visible === false &&
    this.props.onTimeout &&
    this.props.onTimeout()

  render() {
    const { color = 'primary', onTimeout, ...props } = this.props

    return (
      <Theme>
        {({ colorize }) => (
          <Motion
            defaultStyle={{ position: -50 }}
            onRest={this.handleRest}
            style={{
              position: spring(this.state.visible ? 50 : -100, {
                stiffness: 180,
                damping: 12,
              }),
            }}
          >
            {style => (
              <View {...props}>
                <View
                  {...styles.bubble}
                  style={{
                    backgroundColor: colorize(color),
                    bottom: style.position,
                  }}
                  direction="row"
                  alignV="center"
                >
                  <View flex={70} direction="row" {...styles.text}>
                    <Text
                      color="textOnBackground"
                      align="center"
                      autoBreak
                      style={{ width: '100%' }}
                    >
                      {this.props.children}
                    </Text>
                  </View>
                  <View
                    flex={30}
                    alignH="center"
                    alignV="center"
                    direction="row"
                  >
                    <Icon name="check-filled" color="white" />
                  </View>
                </View>
              </View>
            )}
          </Motion>
        )}
      </Theme>
    )
  }
}

export default NotificationBubble
