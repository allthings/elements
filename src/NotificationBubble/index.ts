export { default } from './NotificationBubble'
export * from './NotificationBubble'
