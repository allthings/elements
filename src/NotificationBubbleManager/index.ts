export { default } from './NotificationBubbleManager'
export * from './NotificationBubbleManager'
