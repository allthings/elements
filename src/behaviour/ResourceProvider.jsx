import React from 'react'
import PropTypes from 'prop-types'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
class ResourceProvider extends React.Component {
  getChildContext() {
    return { resourcePath: this.props.resourcePath }
  }

  static childContextTypes = {
    resourcePath: PropTypes.string,
  }

  static propTypes = {
    resourcePath: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  }

  static defaultProps = {
    resourcePath: RESOURCE_PATH,
    children: null,
  }

  render() {
    return this.props.children
  }
}

export default ResourceProvider
