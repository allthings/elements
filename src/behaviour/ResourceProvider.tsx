import React from 'react'

export const RESOURCE_PATH = 'https://static.allthings.me'

/**
 * Elements uses a set of static resources like images or icons.
 * In order to benefit from caching across all apps, these resources are provided by a static asset CDN.
 *
 * Whenever you like to use Icons or Illustrations, you need to use the ResouceProvider to let the components know where they are.
 */
const ResouceProvider = React.createContext({
  resourcePath: RESOURCE_PATH,
})

export type ResourceType = {
  resourcePath?: string
}

export default ResouceProvider
