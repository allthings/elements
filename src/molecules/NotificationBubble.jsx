import React from 'react'
import View from '../atoms/View'
import { css } from 'glamor'
import Theme from '../behaviour/Theme'
import Icon from '../atoms/Icon'
import Text from '../atoms/Text'
import { Motion, spring } from 'react-motion'
import PropTypes from 'prop-types'

const styles = {
  bubble: css({
    zIndex: 200,
    position: 'fixed',
    bottom: 50,
    left: '50%',
    transform: 'translate(-50%, 0)',
    margin: '0 auto',
    width: 260,
    height: 68,
    borderRadius: 45,
    boxShadow: '2px 2px 14px 1px rgba(35,35,35,0.2)',
  }),
  text: css({
    width: 200,
    marginLeft: 20,
  }),
}

class NotificationBubble extends React.Component {
  static propTypes = {
    color: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    onTimeout: PropTypes.func.isRequired,
  }

  state = {
    visible: true,
  }

  static defaultProps = {
    color: 'primary',
    onTimeout: _ => _,
  }

  componentDidMount() {
    setTimeout(() => this.setState({ visible: false }), 2000)
  }

  handleRest = () => this.state.visible === false && this.props.onTimeout()

  render() {
    // eslint-disable-next-line no-unused-vars
    const { color, ...props } = this.props
    return (
      <Theme>
        {({ colorize }) => (
          <Motion
            defaultStyle={{ position: -50 }}
            onRest={this.handleRest}
            style={{
              position: spring(this.state.visible ? 50 : -100, {
                stiffness: 180,
                damping: 12,
              }),
            }}
          >
            {style => (
              <View {...styles.container}>
                <View
                  {...styles.bubble}
                  style={{backgroundColor: colorize(color), bottom: style.position}}
                  direction="row"
                  alignV="center"
                >
                  <View flex={70} direction="row" {...styles.text}>
                    <Text
                      color="textOnBackground"
                      align="center"
                      autoBreak
                      style={{ width: '100%' }}
                    >
                      {this.props.children}
                    </Text>
                  </View>
                  <View
                    flex={30}
                    alignH="center"
                    alignV="center"
                    direction="row"
                  >
                    <Icon name="check-filled" color="white" />
                  </View>
                </View>
              </View>
            )}
          </Motion>
        )}
      </Theme>
    )
  }
}

export default NotificationBubble
