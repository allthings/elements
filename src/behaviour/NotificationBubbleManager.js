import React from 'react'
import View from '../atoms/View'
import mitt from 'mitt'
import NotificationBubble from '../molecules/NotificationBubble'
import PropTypes from 'prop-types'
import Message from './Message'

const emitter = mitt()

const send = (message, type) => {
  emitter.emit(type, message)
}

export const sendSuccess = message => send(message, 'success')
export const sendWarning = message => send(message, 'warning')
export const sendInfo = message => send(message, 'info')

class NotificationBubbleManager extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    renderBubble: PropTypes.func,
  }

  state = {
    messages: [],
  }

  componentDidMount() {
    emitter.on('*', this.handleEvent)
  }

  componentWillUnmount() {
    emitter.off('*', this.handleEvent)
  }

  handleEvent = (type, message) =>
    this.setState(({ messages }) => ({
      messages: [...messages, new Message(type, message)],
    }))

  handleTimeout = () =>
    this.setState(({ messages }) => ({ messages: messages.slice(1) }))

  renderBubble = message =>
    this.props.renderBubble ? (
      this.props.renderBubble({
        key: message.id,
        onTimeout: this.handleTimeout,
        children: message.text,
      })
    ) : (
      <NotificationBubble key={message.id} onTimeout={this.handleTimeout}>
        {message.text}
      </NotificationBubble>
    )

  render() {
    const { children, renderBubble, ...props } = this.props
    const message = this.state.messages[0]

    return (
      <View {...props}>
        {message && this.renderBubble(message)}
        {children}
      </View>
    )
  }
}

export default NotificationBubbleManager
