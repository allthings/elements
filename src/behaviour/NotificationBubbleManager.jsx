import React from 'react'
import View from '../atoms/View'
import mitt from 'mitt'
import NotificationBubble from '../molecules/NotificationBubble'
import PropTypes from 'prop-types'

const emitter = mitt()

const send = (message, type) => {
  emitter.emit(type, message)
}

export const sendSuccess = message => send(message, 'success')
export const sendWarning = message => send(message, 'warning')
export const sendInfo = message => send(message, 'info')

class NotificationBubbleManager extends React.Component {
  static propTypes = {
    children: PropTypes.node,
  }

  state = {
    messages: [],
  }

  messageCounter = 0

  componentDidMount() {
    emitter.on('*', this.handleEvent)
  }

  componentWillUnmount() {
    emitter.off('*', this.handleEvent)
  }

  handleEvent = (type, message) =>
    this.setState(({ messages }) => ({
      messages: [
        ...messages,
        {
          type,
          id: this.messageCounter++,
          text: message,
          date: new Date(),
        },
      ],
    }))

  handleTimeout = () => {
    this.setState(({ messages }) => ({
      messages: messages.slice(1),
    }))
  }

  render() {
    const message = this.state.messages[0]
    return (
      <View>
        {message && (
          <NotificationBubble key={message.id} onTimeout={this.handleTimeout}>
            {message.text}
          </NotificationBubble>
        )}
        {this.props.children}
      </View>
    )
  }
}

export default NotificationBubbleManager
