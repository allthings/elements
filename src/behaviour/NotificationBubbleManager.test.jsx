import React from 'react'
import renderer from 'react-test-renderer'
import NotificationBubbleManager, {
  sendSuccess,
} from './NotificationBubbleManager'
import NotificationBubble from '../molecules/NotificationBubble'
import ResourceProvider from '../behaviour/ResourceProvider'

test('NotificationBubbleManager renders the bubble', () => {
  const renderBubble = jest.fn(props => <NotificationBubble {...props} />)

  const nbm = (
    <ResourceProvider>
      <NotificationBubbleManager renderBubble={renderBubble}>
        <div>Test</div>
      </NotificationBubbleManager>
    </ResourceProvider>
  )

  const tree = renderer.create(nbm)
  sendSuccess('success')
  tree.update(nbm)
  expect(renderBubble).toBeCalled()
  expect(tree).toMatchSnapshot()
})
