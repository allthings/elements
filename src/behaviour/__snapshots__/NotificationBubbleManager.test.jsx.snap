// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBubbleManager renders the bubble 1`] = `
.css-1uw148h,
[data-css-1uw148h] {
  z-index: 200;
  position: fixed;
  bottom: 50px;
  left: 50%;
  transform: translate(-50%, 0);
  margin: 0 auto;
  width: 260px;
  height: 68px;
  border-radius: 45px;
  box-shadow: 2px 2px 14px 1px rgba(35,35,35,0.2);
  -webkit-transform: translate(-50%, 0);
}

.css-13ph0e0,
[data-css-13ph0e0] {
  width: 200px;
  margin-left: 20px;
}

.css-4p3ux5,
[data-css-4p3ux5] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-h7582n,
[data-css-h7582n] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 0 0 auto;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 0 0 auto;
}

.css-ms4fmp,
[data-css-ms4fmp] {
  box-sizing: border-box;
  align-content: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  flex: 1 1 70%;
  -webkit-align-content: stretch;
  -webkit-box-align: stretch;
  -webkit-align-items: stretch;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -webkit-flex: 1 1 70%;
}

.css-16gi3pi,
[data-css-16gi3pi] {
  display: block;
  font-family: "Open Sans", Helvetica, Arial, sans-serif;
  font-size: 14px;
  text-align: center;
  white-space: pre-wrap;
  word-break: break-word;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: #ffffff;
}

.css-17pfsf1,
[data-css-17pfsf1] {
  box-sizing: border-box;
  align-content: center;
  align-items: center;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: row;
  justify-content: center;
  flex: 1 1 30%;
  -webkit-align-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -webkit-flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-flex: 1 1 30%;
}

<div
  data-css-4p3ux5=""
>
  <div
    data-css-4p3ux5=""
  >
    <div
      data-css-1uw148h=""
      data-css-h7582n=""
      style={
        Object {
          "backgroundColor": "#3598db",
          "bottom": -50,
        }
      }
    >
      <div
        data-css-13ph0e0=""
        data-css-ms4fmp=""
      >
        <div
          data-css-16gi3pi=""
          data-css-4p3ux5=""
          style={
            Object {
              "width": "100%",
            }
          }
        >
          success
        </div>
      </div>
      <div
        data-css-17pfsf1=""
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          data-css-4p3ux5=""
          size={32.5}
          style={
            Object {
              "fill": "white",
              "height": 32.5,
              "stroke": false,
              "width": 32.5,
            }
          }
        />
      </div>
    </div>
  </div>
  <div>
    Test
  </div>
</div>
`;
